`storage`: add self compaction counter tunable for use in `adjacent_merge_compact()`

WillemKauf opened the pull request:
In order to allow adjacent merge compaction to make forward progress in all situations, change this unconditional `co_return;` after self compaction to a conditional `break;` or `co_return;` depending on the number of segments that have been self compacted since the last adjacent merge.

Having this as a tunable allows control over the read/write amplification of adjacent merge compaction when sliding window compaction is disabled.

Tagged for backports.

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v24.3.x
- [X] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* Adds the tunable `log_compaction_adjacent_merge_self_compaction_count`, which allows for adjacent merge compaction to make more forward progress in all situations when `log_compaction_use_sliding_window` is disabled.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--1cbd7a61-59c3-4532-9066-4fcfdea18a8b-->
<!--TEST RESULTS-->
<!--3536764c-5ec0-4e72-8a95-4d1c297b86e0-->
<!--TEST RESULTS-->
<!--0b72656a-8b7a-4a5c-aa4e-aeca3087137f-->
<!--TEST RESULTS-->
<!--bd5240fe-f6f6-4c4a-9b5c-ed753e0ca38b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62038">build#62038</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_scrubber_test.CloudStorageScrubberTest.test_scrubber.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62038#0195205d-7ec2-4e69-94f8-4fa25be2ee6c | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/62038#0195205d-7ec1-47e1-a291-14d8dcb3cb0d | FLAKY | 1/3 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/62038#01952060-7ae1-440b-afd1-08aeaf1f3a3e | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62038#01952060-7ae0-4094-80d7-f5bfc37a611d | FLAKY | 1/3 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62038#01952060-7ae1-440b-afd1-08aeaf1f3a3e | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62038#01952060-7ae0-4094-80d7-f5bfc37a611d | FLAKY | 1/3 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_relaxed_acks.write_caching=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/62038#01952060-7ade-4fad-95e7-3f85d282e213 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62076">build#62076</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_concurrent_truncations.cloud_storage_enabled=True.truncate_point=random_offset | ducktape | https://buildkite.com/redpanda/redpanda/builds/62076#0195258b-4658-4214-ad86-8736b1cf1fab | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62076#0195258f-d19e-47d3-9369-93ab4c3327cd | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62076#0195258f-d19e-47d3-9369-93ab4c3327cd | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62273">build#62273</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_concurrent_truncations.cloud_storage_enabled=True.truncate_point=random_offset | ducktape | https://buildkite.com/redpanda/redpanda/builds/62273#01953e11-55ca-4396-aa37-48d7db847ca4 | FLAKY | 1/2 |
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_partition_move_x_core.replication_factor=1.unclean_abort=False.recovery=restart_recovery.compacted=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/62273#01953e11-55c8-4c79-9356-06f51792f84a | FLAKY | 1/2 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsets.test_consumer_group_offsets | ducktape | https://buildkite.com/redpanda/redpanda/builds/62273#01953e11-55c8-4c79-9356-06f51792f84a | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62307">build#62307</a></summary>


| test_id                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62307#01953f75-8258-4198-b898-87039388e70d | FLAKY         | 1/2      |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_cross_shard.num_to_upgrade=2.cloud_storage_type=CloudStorageType.ABS     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62307#01953f75-8258-4198-b898-87039388e70d | FLAKY         | 1/2      |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62307#01953f75-8259-4ee4-bacd-dd505d10a751 | FLAKY         | 1/2      |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit                                                                          | unit        | https://buildkite.com/redpanda/redpanda/builds/62307#01953f2e-0238-4c8a-9d75-dedf63814c3a | FLAKY         | 1/2      |
</details>
<!--/body-->

WillemKauf commented:
Force push to:
* Rebase to `upstream/dev` to fix merge conflicts

dotnwat reviewed (commented):
>  we should not need to self compact every segment in the log before attempting adjacent compaction.

functionality the change looks fine. but what's the motivation? i'm definitely not arguing for the status quo, but "shouldn't need to self-compact every segment" doesn't seem obvious to me as motivation for the change. can you elaborate a bit on that?

WillemKauf commented:
>functionality the change looks fine. but what's the motivation? i'm definitely not arguing for the status quo, but "shouldn't need to self-compact every segment" doesn't seem obvious to me as motivation for the change. can you elaborate a bit on that?

This particular code path is taken when `log_compaction_use_sliding_window` is `false` at the cluster level- so imagine this is now the "primary" code path for compaction/housekeeping of a log.

It seems entirely feasible that for high ingress work loads, the rate at which segments are rolled in a log could very easily exceed the interval bound set by `log_compaction_interval_ms`.

With the current behavior in this high ingress situation, we would NEVER see any adjacent compaction occur, since we have to first self compact every segment in the log. In this situation, we don't have sliding window compaction active, and we effectively don't have adjacent segment compaction active either. We are limited to self compaction of each segment in the log, meaning the number of segments in the log could continue to grow unbounded (bar other retention policies) and on top of that, compaction of the key set could be ineffective/essentially zero if it spans segment boundaries.

This seems like undesired behavior to me. I think we want to give adjacent compaction a chance to make forward progress if it can. That is the motivation behind this change.

Let me know your thoughts on this.

dotnwat commented:
> With the current behavior in this high ingress situation, we would NEVER see any adjacent compaction occur

thanks! but is this true? i thought that dynamic scheduling group changes to the compaction fiber would increase compaction CPU share as the backlog grew, so that if ingest outpaced compaction eventually compaction would be a favored?

WillemKauf commented:
>thanks! but is this true? i thought that dynamic scheduling group changes to the compaction fiber would increase compaction CPU share as the backlog grew, so that if ingest outpaced compaction eventually compaction would be a favored?

Potentially the PID controller would help with providing compaction with more scheduling group shares if the fibers were competing, but I still believe that ultimately, the behavior for this situation is dictated by the ingress rate (effectively, the segment roll interval) vs the `log_compaction_interval_ms`.

dotnwat commented:
>I believe ultimately the behavior for this situation is dictated by the ingress rate (effectively, the segment roll interval) vs the log_compaction_interval_ms

but isn't the situation you describe ultimately about the compaction backlog, which the PID controller does monitor?

The other question I had was related to the cost here. Prior to this we would compact all the segments then do adjacent segment merging. If I understand correctly it would be like this, for compacted segment A and uncompacted segments BCD

A'BCD -> A'B'C'D' -> A''

where A'' would be the result of adjacent segment compaction and folding right. In particular, we'd realize the space savings of BCD prior to adjacent segment compaction.

Now with this change it looks like we do the following?

A'BCD -> A'B'CD -> A''CD then,
A''CD -> A''C'D -> A'''D then,
A'''D -> A'''D' -> A''''

Maybe these end up being equivalent in the end, but I'm wondering about the cost of adajcent segment compaction. If A' is very large, it seems like we need to do a lot more work (but maybe work here just translates into time) before we realize teh space savings of compacting BCD, where as in the current system would we realize the space savings of compacting BCD prior to incurring the potentially time cost of adjacent segment compaction of A' is large?

Ultimately the answer may be that we don't want a granularity of 1 nor of N, but something in between?

WillemKauf commented:
> but isn't the situation you describe ultimately about the compaction backlog, which the PID controller does monitor?

I understand that the controller does prescribe scheduling group shares, which has the effect of giving certain fibres CPU priority over other fibres in a different scheduling group, but I question if this actually has any effect over the situation I'm describing. 

I mean, we're still sleeping for `log_compaction_interval_ms`, right? We're not putting the produce path on hold for that time length, there is still ingress, segments are still being rolled. The controller does help compaction make progress by giving it priority if it were competing for CPU concurrently with other fibres, but as far as I can picture it, it's not saving us from the situation here.

Please correct me if anything I wrote above seems incorrect.

> A'BCD -> A'B'C'D' -> A''
> where A'' would be the result of adjacent segment compaction and folding right. In particular, we'd realize the space savings of BCD prior to adjacent segment compaction.

Adjacent segment compaction, at time of writing, is still subject to the constraint of only compacting two (self-compacted) segments at once (I am working on this atm too :smile:)

So, in both situations, whether `co_return;` or `break;`, the scenario you laid out actually plays out the same:

Abridged steps: 

```
A'BCD -> A'B'C'D'
A'B'C'D' -> A"C'D'
A"C'D' -> A"C"
A"C" -> A""
```

where `A"` is the first result of adjacent compaction merging, `A""` is the second result of adjacent compaction merging of segment `A`. 

The difference is how quickly we get there. In the current behavior, the steps (un-abridged) would be:

```
A'BCD -> A'B'CD
A'B'CD -> A'B'C'D
A'B'C'D -> A'B'C'D'
A'B'C'D' -> A"C'D'
A"C'D' -> A"C"
A"C" -> A""
```

with the proposed changes (un-abridged steps), we get there faster:

```
A'BCD -> A'B'CD
A'B'CD -> A"C'D
A"C'D -> A"C'D'
A"C'D' -> A"C"
A"C" -> A""
```


You raise a good point about what would happen to the cost if we were able to concatenate `n > 2` segments at a time. However, once we have that utility, I think the procedure for adjacent segment compaction would look fairly different (we would probably want to do something similar to sliding window in that every segment is self compacted before doing any other work. Then, we perform concatenation of all compatible sub-arrays of `m_i` segments over the entire group `n`, and then self compact the replacement segments of each sub-array).

dotnwat commented:
> The difference is how quickly we get there
> with the proposed changes (un-abridged steps), we get there faster:

But is "there" the goal (I'm assuming "there" means the final state)?

Say A' is large and it takes a long time to compact/combine A' and its neighbor (say COST_1) to the right because A' is large (so COST_1 is the dominant factor).

So if we have A'BCD, then currently we self compact BCD -> B'C'D' which happens before self compaction. Self compacting B'C'D' takes COST_2 = COST_B+C+D.

Now today then we have

COST_2 gets us into the state A'B'C'D' which may have good space savings, for example, if B'C'D' compact well on their own in isolation.

But with the change in this PR wouldn't we have

COST_B + COST_A gets us to the state A''CD (where self-compacted B has been absorbed into A')

then we'd have

COST_C + COST_A (again) gets us to the state A'''D ...

but of course its very complicated because maybe compaction ratios are amazing when the combinations happen, or the costs are backwards, or they are a wash.

I'm just having a hard to time understanding if we are always doing better with this change, the same in the worst case, or sometimes worse...etc?

WillemKauf commented:
Right, I guess I should mention my steps were making some incorrect assumptions which would not necessarily be true (A"' could indeed be the result of combining A" and C', instead of making C" from C' and D'...)

> I'm just having a hard to time understanding if we are always doing better with this change, the same in the worst case, or sometimes worse...etc?

I think we could draw out situations where either approach is better or worse. I think it is a wash, incredibly dependent on the distribution of the key set across segment boundaries. Once again, the main objective of this change is being able to make forward progress with adjacent segment compaction in any situation.

dotnwat reviewed (dismissed):
cool lets see how it goes

WillemKauf commented:
>cool lets see how it goes

Feelings on backports?

WillemKauf committed:
`config`: add `log_compaction_adjacent_merge_self_compaction_count`

The number of self compactions that must occur before an adjacent compaction
is attempted in the log. If set to `std::nullopt`, every segment in the log
must be self-compacted before an adjacent compaction is attempted.

WillemKauf committed:
`storage`: use cluster tunable in `adjacent_merge_compact()`

In order to allow adjacent merge compaction to make forward progress
in all situations, change this unconditional `co_return;` after
self compaction to a conditional `break;` or `co_return;` depending
on the number of segments that have been self compacted since the last
adjacent merge.

Having this as a tunable allows control over the read/write amplification
of adjacent merge compaction when sliding window compaction is disabled.

WillemKauf commented:
Force push to:
* Add tunable `log_compaction_adjacent_merge_self_compaction_count` and use it in `disk_log_impl::adjacent_merge_compact()`.

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -516,7 +516,21 @@ ss::future<> disk_log_impl::adjacent_merge_compact(
             subtract_segment_bytes(segment, removed_bytes);
-            co_return;
+
+            auto& config_adjacent_merge_count
```
why a mutable reference?

WillemKauf replied:
No reason. Removed

WillemKauf commented:
Force push to:
* Remove `&` in definition of `config_adjacent_merge_count` variable

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25119-v24.3.x-846 remotes/upstream/v24.3.x
git cherry-pick -x 52dc680259 be438c2a34
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/13535036078).

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25119-v24.2.x-560 remotes/upstream/v24.2.x
git cherry-pick -x 52dc680259 be438c2a34
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/13535036271).

dotnwat commented:
whats the logic on backporting?

WillemKauf commented:
>whats the logic on backporting?

If we get into situations where we end up disabling `log_compaction_use_sliding_window` (a la recent events), this PR would allow us even finer tuning over adjacent segment compaction.

It's not a bug fix but it does seem beneficial to have in `24.3` and `24.2` at least. WDYT?

WillemKauf commented:
@dotnwat is the thumbs up emoji an approval for fulfilling the backports?

dotnwat commented:
> @dotnwat is the thumbs up emoji an approval for fulfilling the backports?

someone mentioned to me yesterday in the context of another PR that they thought there were issues with backports that introduce new configuration options. maybe that's not true tho. i think we need to confirm that.

WillemKauf commented:
The backport foot gun reportedly would happen in a situation where, e.g., we backport a cluster config to `v24.2.x` but not `v24.3.x`, and upon upgrade from `v24.2.x` to `v24.3.x`, we run into issues.

This is tagged for backport to both `v24.2.x` and `v24.3.x`, so it would be safe if we backport in descending version order AND we also would need to cut releases in the same order. 

Let me know if you'd rather just avoid backporting in this situation completely, sounds like the risk isn't negligible.

dotnwat commented:
> Let me know if you'd rather just avoid backporting in this situation completely, sounds like the risk isn't negligible.

@WillemKauf Thanks I learned something new with that slack thread on backporting. Yeh it doesn't seem worth the effort to coordinate backports for this change. The edge cases where we end up with a problematically large number of small segments which IIUC this would address is pretty rare.
